runtime/compiler/codegen/J9AheadOfTimeCompile.cpp

```diff
@@ -36,20 +36,97 @@
 #include "runtime/RelocationRuntime.hpp"
 #include "runtime/RelocationRecord.hpp"
 #include "runtime/SymbolValidationManager.hpp"
+#if defined(J9VM_OPT_JITSERVER)
+#include "runtime/JITClientSession.hpp"
+#endif /* defined(J9VM_OPT_JITSERVER) */
+
 
 extern bool isOrderedPair(uint8_t reloType);
 
 uintptr_t
-J9::AheadOfTimeCompile::getClassChainOffset(TR_OpaqueClassBlock *classToRemember)
+J9::AheadOfTimeCompile::getClassChainOffset(TR_OpaqueClassBlock *classToRemember,
+                                            const AOTCacheClassChainRecord *&classChainRecord)
+   {
+   TR_J9VMBase *fej9 = (TR_J9VMBase *)self()->comp()->fe();
+   TR_SharedCache *sharedCache = fej9->sharedCache();
+   void *classChain = sharedCache->rememberClass(classToRemember, &classChainRecord);
+   if (!classChain)
+      self()->comp()->failCompilation<J9::ClassChainPersistenceFailure>("classChain == NULL");
+   return self()->offsetInSharedCacheFromPointer(sharedCache, classChain);
+   }
+
+#if defined(J9VM_OPT_JITSERVER)
+
+void
+J9::AheadOfTimeCompile::addClassSerializationRecord(const AOTCacheClassChainRecord *classChainRecord,
+                                                    const uintptr_t *romClassOffsetAddr)
+   {
+   const AOTCacheClassRecord *record = classChainRecord ? classChainRecord->records()[0] : NULL;
```

Comment:
Could you explain how this is all laid out? `classChainRecord` is of type 
```
class AOTCacheClassChainRecord final : public AOTCacheListRecord<ClassChainSerializationRecord, AOTCacheClassRecord>
```
which means 
```const R *const *records() const { return (const R *const *)_data.end(); }``` 
becomes 
```const AOTCacheClassRecord *const *records() const { return (const AOTCacheClassRecord *const *)_data.end(); }```
where `D _data` is now `ClassChainSerializationRecord _data`. 

Searching for `end()` takes me to `AOTSerializationRecord::end()`, which is
```
const uint8_t *end() const { return (const uint8_t *)this + size(); }
```
I don't really understand how the data is laid out, because at first glance this looks like we're going out of bounds.. I think this needs a bunch of documentation to describe how all of this fits together,

Reply:
The layout of `AOTCacheClassChainRecord` is as follows:
- `AOTCacheClassChainRecord` header
- `ClassChainSerializationRecord` header
- `ClassSerializationRecord` data: array of N class IDs
- array of N `AOTCacheClassRecord` pointers

I'll add comments in `JITServerAOTCache.hpp` to better describe the layout. I'll also move `classChainRecord->records()[0]` to a separate helper method in `AOTCacheClassChainRecord`.